### ISSUE TITLE
Are permissions applied to Fargate containers broken?

### ISSUE THREAD

**Issue by mattfysh** [aws-amplify/amplify-cli#8427]
>**Note**: If your question is regarding the AWS Amplify Console service, please log it in the 
[AWS Amplify Console repository](https://github.com/aws-amplify/amplify-console/issues)

**Which Category is your question related to?**

ECS / Fargate

**Amplify CLI Version** 

You can use `amplify -v` to check the amplify cli version on your system

v6.2.1

**What AWS Services are you utilizing?**

ECS / Fargate

**Provide additional details e.g. code snippets**

I'm seeing some very strange behaviour with containers + permissions applied via `amplify api update`

1. After going through the CLI prompts for `amplify api update` and adding permissions to access some auth + storage resources, I am not seeing any changes made to git-tracked files
2. I have to run `amplify push`, and I can see the changes I expected to see in Step 1
3. When a teammate pulls and then pushes the same environment, the permissions added in Step 1 are lost again

Are there any known issues with permissions and containers?

**Comment by mattfysh** [comment]
Another strange thing that's happening is that two nearly identical task definitions are being created:

(a) One by `amplify push`
(b) And the other by one of the stages in CodePipeline

I've run a diff on the JSON definitions for each task, and the only differences are:

1. `.revision` - expected
2. `.containers[0].image` - the definition created by (a) contains `:latest` tag on the end, the definition created by (b) doesn't

**Comment by edwardfoyle** [comment]
Hi @mattfysh, when updating the resource access permissions on your containers resource you should see changes to the containers CloudFormation template as well as changes in the `backend-config.json` which should both be git-tracked files. Are you seeing changes to those files after updating the access permissions?

**Comment by mattfysh** [comment]
What I am seeing doesn't fill me with confidence. I am trying to do everything slowly and document the steps, here's what I have so far:

## Resetting the repo

In the following steps I reset the environment a couple of times, this is done using:

```
# simulate a freshly cloned repo
git reset --hard
git clean -dffx
amplify init
amplify env pull # probably redundant but did it anyway
```

## Part I - Remove all but one permission

1. Reset repo
2. `amplify api update` - remove all but one permission via CLI. I do want to reset permissions to a clean state, but I don't trust that answering "No" for "Do you want to access other resources in this project from your api?" will work correctly
3. After removing all but one permission, the only file that changed is `amplify/backend/backend-config.json` : a few `api.dependsOn` entries were removed, but no other git-tracked file has been updated
4. Importantly - the CF template for the container has NOT been updated, the permissions that were just removed via the CLI are still present
5. Change some of my JS code to force the container to redeploy
6. Run `amplify push`, just after the deployment starts I can see that the CF template file has been updated to remove the permissions. Only the one permission remains
7. The CF deployment ultimately fails because an environment variable used for a removed permission was not removed from the task definition (the Ref is no longer a parameter)
8. Nothing in the CLI would remove this so I remove it manually and try to push again
9. The variable removed from the CF template has been restored, I then find it is also in backend-config.json and remove it from both files, try to push again - but same result
10. Eventually I find the variable still exists in `amplify-meta.json` which is not under git version control. I remove it there and everything is working as expected - once deployment has completed I confirm the cloud environment also looks as expected
11. git commit and push changes to files

## Part II - Restore one of the removed permissions

1. Reset repo (to simulate a teammate carrying out this work on a fresh repo)
2. `amplify api update` - okay that's **very odd** - the preselected options don't reflect all of the work in Part I 😢 
3. CTRL+C to abort the CLI
4. Make a small JS change to force the container to redeploy
5. The CF template is updated to restore ALL OF THE PERMISSIONS removed in Part I
6. The push fails anyway, because the permissions statement contains Refs that don't exist in Parameters

So there are A LOT of issues at play here, and I don't understand if the CLI is broken or I'm doing something wrong. The only path forward I can think of is to delete and recreate the resource, there is some kind of state which seems to be stuck somewhere which is messing up the standard git workflow with containers

Also don't forget the other issue I mentioned above which is the push process creating two identical task definitions - when one would suffice.

**Comment by mattfysh** [comment]
Ok, traced it back to `api.<name>.categoryPolicies`

The only file that has the correct data after updating permissions via CLI is `#current-cloud-backend/amplify-meta.json`

These three files do not have the right information:

• `#current-cloud-backend/backend-config.json`
• `backend/backend-config.json`
• `backend/amplify-meta.json`

**Comment by mattfysh** [comment]
Ok think I've fixed it.

1. In `backend/backend-config.json` I've reset these properties on `api.<name>`:

```
"environmentMap": {},
"categoryPolicies": [],
"mutableParametersState": {},
```

2. Run `amplify env checkout <current_env>`
3. Run `amplify api update`, no permissions preselected, applied all required permissions
4. Run `amplify push`

Now when I check backend-config.json in both local and cloud - those 3 properties are empty objects/arrays

And when I check amplify-meta.json in both local and cloud - those 3 properties are wholly and correctly populated

Not sure how I ended up down this rabbit hole or how the amplify files got into such a state, but hopefully there is something that can be fixed in the CLI to prevent developers from having to investigate Amplify internals -- it's been a long week!
